ouroboros-network-framework/test/Test/Simulation/Network/Snocket.hs

```diff
@@ -627,28 +636,202 @@ prop_generator_NonFailingBeararInfoScript (NonFailingBearerInfoScript s) = not (
+            _ <- wait serverAsync
+            return res
+
+prop_connect_and_not_close :: BearerInfoScript -> Property
+prop_connect_and_not_close (BearerInfoScript script) =
+  prop_connect_to_template sim
+  where
+    sim :: forall s . IOSim s (Either SomeException ())
+    sim =
+      withSnocket nullTracer (toBearerInfo <$> script) (\snocket ->
+        withAsync
+          (server (TestAddress (0 :: Int)) snocket (\_ _ -> pure ()) accept')
+          $ \serverAsync -> do
+            res <- client (TestAddress 1) (TestAddress 0) snocket (\_ _ -> pure ())
+            _ <- wait serverAsync
+            return res
+                                                       )
+      `catch` (\(err :: SomeException) ->
+        -- Should error with NotReleasedListeningSockets
+        case (fromException err
+               :: Maybe (ResourceException (TestAddress Int))) of
+          Just _ ->
+            return (Right ())
+          Nothing ->
+            return (Left err)
+      )
+
+prop_connect_to_not_accepting_socket :: BearerInfoScript -> Property
+prop_connect_to_not_accepting_socket (BearerInfoScript script) =
+  prop_connect_to_template sim
+  where
+    sim :: forall s . IOSim s (Either SomeException ())
+    sim =
+      withSnocket nullTracer (toBearerInfo <$> script) $ \snocket ->
+        withAsync (server (TestAddress (0 :: Int)) snocket close loop)
+          $ \_ -> do
+            res <- client (TestAddress 1) (TestAddress 0) snocket close
+            case res of
+              -- Should timeout
+              Left _ -> return (Right ())
+              Right _ -> return (Left (toException UnexpectedOutcome))
+
+prop_connect_to_uninitialised_socket :: BearerInfoScript -> Property
+prop_connect_to_uninitialised_socket (BearerInfoScript script) =
+  prop_connect_to_template sim
+  where
+    sim :: forall s . IOSim s (Either SomeException ())
+    sim =
+      withSnocket nullTracer (toBearerInfo <$> script) $ \snocket -> do
+        res <- client (TestAddress (1 :: Int)) (TestAddress 0) snocket close
+        case res of
+          -- Should complain about no such listening socket
+          Left _ -> return (Right ())
+          Right _ -> return (Left (toException UnexpectedOutcome))
+
+prop_simultaneous_open :: BearerInfoScript -> Property
+prop_simultaneous_open (BearerInfoScript script) =
+  prop_connect_to_template sim
+  where
+    sim :: forall s . IOSim s (Either SomeException ())
+    sim =
+      withSnocket nullTracer (toBearerInfo <$> script) $ \snocket -> do
+        withAsync
+          (connectingServer (TestAddress (0 :: Int)) (TestAddress 1) snocket)
+          $ \clientAsync -> do
+            _ <- connectingServer (TestAddress 1) (TestAddress 0) snocket
+            wait clientAsync
 
 --
 -- Utils
 --
 
+server
+  :: ( MonadThread m
+    , MonadThrow m
+    )
+  => TestAddress addr
+  -> Snocket m fd (TestAddress addr)
+  -> (Snocket m fd (TestAddress addr) -> fd -> m ())
```

Comment:
Let's add a haddock.

We dont need to pass `Snocket` to it, it is the same as the one passed in the previous argument.